neural_speed/core/ne_layers.c

```diff
@@ -6889,157 +6959,809 @@ static void ne_compute_forward_mul_mat_bias_q_f32_bestla(const struct ne_compute
+      const int64_t i2 = i12;
+      const int64_t i3 = i13;
+
+      char* src0_row = (char*)src0_cur->data;
+      char* src1_col = (char*)src1->data + (i11 * nb11 + i12 * nb12 + i13 * nb13);
+
+      float* dst_col = (float*)((char*)dst->data + (i1 * nb1 + i2 * nb2 + i3 * nb3));
+
+      // parallelize by src0 rows
+
+      bestla_f32f32_forward((float*)src1_col, (float*)src0_row, dst_col, 1, ne0, ne10, nb11 / ne_element_size(src1),
+                            nb1 / ne_element_size(dst), params->wdata);
+    }
+  }
+}
+static void ne_compute_forward_mul_mat_id(const struct ne_compute_params* params, const struct ne_tensor* ids,
+                                          const struct ne_tensor* src1, struct ne_tensor* dst) {
+  switch (dst->opt[0]->type) {
+    case NE_TYPE_Q4_0:
+    case NE_TYPE_Q4_1:
+    case NE_TYPE_Q5_0:
+    case NE_TYPE_Q5_1:
+    case NE_TYPE_Q8_0:
+    case NE_TYPE_Q6_K:
+    case NE_TYPE_Q8_1: {
+      ne_compute_forward_mul_mat_id_q_f32(params, ids, src1, dst);
+    } break;
+    case NE_TYPE_BTLA: {
+      ne_compute_forward_mul_mat_id_q_f32_bestla(params, ids, src1, dst);
+    } break;
+    case NE_TYPE_F16: {
+      ne_compute_forward_mul_mat_id_f16_f32(params, ids, src1, dst);
+    } break;
+    case NE_TYPE_F32: {
+      ne_compute_forward_mul_mat_id_f32(params, ids, src1, dst);
+    } break;
+    default: {
+      NE_ASSERT(false);
+    } break;
+  }
+}
+static void ne_compute_forward_argsort_f32(const struct ne_compute_params* params, const struct ne_tensor* src0,
+                                           struct ne_tensor* dst) {
+  if (params->type == NE_TASK_INIT || params->type == NE_TASK_FINALIZE) {
+    return;
+  }
+  const int64_t ne00 = src0->ne[0];
+  const int64_t ne01 = src0->ne[1];
+  const int64_t ne02 = src0->ne[2];
+  const int64_t ne03 = src0->ne[3];
+
+  const int64_t ne0 = dst->ne[0];
+  const int64_t ne1 = dst->ne[1];
+  const int64_t ne2 = dst->ne[2];
+  const int64_t ne3 = dst->ne[3];
+
+  const size_t nb00 = src0->nb[0];
+
+  const size_t nb01 = src0->nb[1];
+  const size_t nb02 = src0->nb[2];
+  const size_t nb03 = src0->nb[3];
+
+  const size_t nb0 = dst->nb[0];
+  const size_t nb1 = dst->nb[1];
+  const size_t nb2 = dst->nb[2];
+  const size_t nb3 = dst->nb[3];
+  const int ith = params->ith;
+  const int nth = params->nth;
+
+  const int64_t nr = ne_nrows(src0);
+
+  for (int64_t i = ith; i < nr; i += nth) {
+    int32_t* dst_data = (int32_t*)((char*)dst->data + i * nb1);
+    const float* src_data = (float*)((char*)src0->data + i * nb01);
+
+    for (int64_t j = 0; j < ne0; j++) {
+      dst_data[j] = j;
+    }
+
+    // C doesn't have a functional sort, so we do a bubble sort instead
+    for (int64_t j = 0; j < ne0; j++) {
+      for (int64_t k = j + 1; k < ne0; k++) {
+        if (src_data[dst_data[j]] < src_data[dst_data[k]]) {
+          int32_t tmp = dst_data[j];
+          dst_data[j] = dst_data[k];
+          dst_data[k] = tmp;
+        }
+      }
+    }
+  }
+}
```

Comment:
you can separate this function into another file like GEMM, ```layers/reduce/argsort.cpp``` and call std::sort instead

Reply:
I didn't find gemm, are you talking about conv.h and conv.cpp?

Reply:
yes

Reply:
down